### PR TITLE
[SPARK-26735][SQL] Verify plan integrity for special expressions

### PR DESCRIPTION
**Pull request by rednaxelafx**
## What changes were proposed in this pull request?

Add verification of plan integrity with regards to special expressions being hosted only in supported operators. Specifically:

- `AggregateExpression`: should only be hosted in `Aggregate`, or indirectly in `Window`
- `WindowExpression`: should only be hosted in `Window`
- `Generator`: should only be hosted in `Generate`

This will help us catch errors in future optimizer rules that incorrectly hoist special expression out of their supported operator.

TODO: This PR actually caught a bug in the analyzer in the test case `SPARK-23957 Remove redundant sort from subquery plan(scalar subquery)` in `SubquerySuite`, where a `max()` aggregate function is hosted in a `Sort` operator in the analyzed plan, which is invalid. That test case is disabled in this PR.
SPARK-26741 has been opened to track the fix in the analyzer.

## How was this patch tested?

Added new test case in `OptimizerStructuralIntegrityCheckerSuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
I like this check! It can save us a lot of time when debugging!

**Comment by cloud-fan** [comment]
LGTM except a code style comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -41,7 +41,39 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
   // Check for structural integrity of the plan in test mode. Currently we only check if a plan is
   // still resolved after the execution of each rule.
```
[the remark this answers is not in the document]
Comment (rednaxelafx) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Thanks! Updated

**Comment by rednaxelafx** [comment]
Thanks for your reviews, @cloud-fan and @viirya !

I've updated the PR addressing your comments.

This PR has actually caught a genuine bug in the analyzer in one of the test cases:
For this query in `SubquerySuite`,
```sql
SELECT *
FROM   t1
WHERE  c1 = (SELECT   max(t2.c1)
             FROM     t2
             GROUP BY t2.c1
             HAVING   count(*) >= 1
             ORDER BY max(t2.c1))
```
The analyzer resolves it into:
```
== Analyzed Logical Plan ==
c1: int, c2: int
Project [c1#51, c2#52]
+- Filter (c1#51 = scalar-subquery#109 [])
   :  +- Project [max(c1)#137]
   :     +- Sort [max(c1#28) ASC NULLS FIRST], true                // NOTE HERE!!
   :        +- Project [max(c1)#137, c1#28]
   :           +- Filter (count(1)#139L >= cast(1 as bigint))
   :              +- Aggregate [c1#28], [max(c1#28) AS max(c1)#137, count(1) AS count(1)#139L, c1#28]
   :                 +- SubqueryAlias `t2`
   :                    +- Project [_1#23 AS c1#28, _2#24 AS c2#29]
   :                       +- LocalRelation [_1#23, _2#24]
   +- SubqueryAlias `t1`
      +- Project [_1#46 AS c1#51, _2#47 AS c2#52]
         +- LocalRelation [_1#46, _2#47]
```
... where `Sort [max(c1#28) ASC NULLS FIRST], true` is an example of a `Sort` operator hosting an aggregate expression `max`, incorrectly.

It's somewhat tedious to fix because we need to tweak the order a bit. Working on it.

**Pull request closed without merging**